DependencyInjectionContainer/DIContainer.cs

```diff
@@ -1,35 +1,48 @@
 ﻿using System.Reflection;
+using System.Runtime.CompilerServices;
 using System.Runtime.ExceptionServices;
 using DependencyInjectionContainer.Enums;
 using DependencyInjectionContainer.Exceptions;
 
 namespace DependencyInjectionContainer
 {
-    public sealed class DIContainer
+    public sealed class DIContainer : IDisposable
     {
-        private readonly IEnumerable<Service> registeredServices;
+        internal readonly IEnumerable<Service> registeredServices;
+        ServicesInstanceList servicesInstanceList;
 
         internal DIContainer(IEnumerable<Service> services, DIContainer parent)
         {
+            servicesInstanceList = new ServicesInstanceList();
             registeredServices = services;
             ContainerParent = parent;
+            IsDisposed = false;
         }
 
-        internal DIContainer ContainerParent { get; init; }
+        public bool IsDisposed { get; private set; }
+        internal DIContainer ContainerParent { get; private set; }
 
-        public DIContainerBuilder CreateChildContainer() => new(this);
+        public DIContainerBuilder CreateChildContainer()
+        {
+            ThrowIfDisposed();
+            return new(this);
+        }
 
         public TTypeToResolve Resolve<TTypeToResolve>(ResolveSource resolveType = ResolveSource.Any) where TTypeToResolve : class
         {
+            ThrowIfDisposed();
             return (TTypeToResolve)Resolve(typeof(TTypeToResolve), resolveType);
         }
 
         public IEnumerable<TTypeToResolve> ResolveMany<TTypeToResolve>(ResolveSource resolveSource = ResolveSource.Any) where TTypeToResolve : class
         {
+            ThrowIfDisposed();
+
             switch (resolveSource)
             {
                 case ResolveSource.Any:
-                    return ResolveLocal().Concat(ResolveNonLocal());
+                    IEnumerable<TTypeToResolve> resolvedLocal = ResolveLocal();
+                    return resolvedLocal.Concat(ResolveNonLocalImplementationTypesWhichWereNotResolved(resolvedLocal));
 
                 case ResolveSource.Local:
```

Comment:
Consider materializing requested instances instead of providing non-materialized IEnumerable otherwise on each access to ResolveMany result will be recalculated which forces user to call materialization on each ResolveMany call